php/Blockonomics.php

```diff
@@ -131,59 +131,57 @@ public function check_get_callbacks_response_code($response){
+            $error_str = $this->examine_server_callback_urls($response_body, $wordpress_callback_url, $api_url, $crypto);
+        }
+        return $error_str;
+    }
+
+    // checks each existing xpub callback URL to update and/or use
+    public function examine_server_callback_urls($response_body, $wordpress_callback_url, $api_url, $crypto)
+    {
+        $base_url = preg_replace('/https?:\/\//', '', $api_url);
+        $available_xpub = '';
+        //Go through all xpubs on the server and examine their callback url
+        foreach($response_body as $one_response){
+            $server_callback_url = isset($one_response->callback) ? $one_response->callback : '';
+            $server_base_url = preg_replace('/https?:\/\//', '', $server_callback_url);
+            $xpub = isset($one_response->address) ? $one_response->address : '';
+            // No callback
+            if(!$server_callback_url){
+                $available_xpub = $xpub;
+            // Exact match
+            }else if($server_callback_url == $wordpress_callback_url){
+                return '';
             }
-            $response_callback_without_schema = preg_replace('/https?:\/\//', '', $response_callback);
-            if(!$response_callback || $response_callback == null)
-            {
-                //No callback URL set, set one 
-                $this->update_callback($callback_url, $crypto, $response_address);
-            }
-            elseif($response_callback_without_schema != $callback_url_without_schema)
-            {
-                $base_url = get_bloginfo('wpurl');
-                $base_url = preg_replace('/https?:\/\//', '', $base_url);
-                // Check if only secret differs
-                if(strpos($response_callback, $base_url) !== false)
-                {
-                    //Looks like the user regenrated callback by mistake
-                    //Just force Update_callback on server
-                    $this->update_callback($callback_url, $crypto, $response_address);
-                }
-                else
-                {
-                    $error_str = __("You have an existing callback URL", 'blockonomics-bitcoin-payments');
-                }
-                
+            // Partial Match - Only secret or protocol differ
+            else if(strpos($server_callback_url, $api_url) === 0 || strpos($server_base_url, $base_url) === 0 ){
+                $this->update_callback($wordpress_callback_url, $crypto, $xpub);
+                return '';
             }
         }
-        else 
-        {
-            $error_str = __("You have an existing callback URL", 'blockonomics-bitcoin-payments');
-            // Check if callback url is set
-            foreach ($response_body as $res_obj)
-             if(preg_replace('/https?:\/\//', '', $res_obj->callback) == $callback_url_without_schema)
-                $error_str = "";
-        }  
+        // Use the available xpub
```

Comment:
Test Case 20 will possibly update to the same callback url, which can create problems if the same url matches two xpubs. We should reinclude the $partial_match flag and only update the partial_match after checking all the callback URLs do not contain an exact match.
```
if($partial_match || $available_xpub){
  $update_xpub = $partial_match ? $partial_match : $available_xpub;
  $this->update_callback($wordpress_callback_url, $crypto, $update_xpub);
  return '';
}
```

Reply:
Smart solution, I updated and pushed it, I'll do some testing tomorrow morning

Reply:
Does the algorithm need to account for if there are two partial matches?

Example:
```
server callback 1:  http://shop.home/?wc-api=WC_Gateway_Blockonomics&secret=asdasdkljlkjasd0912312a
server callback 2:  https://shop.home/?wc-api=WC_Gateway_Blockonomics&secret=f5e53a0f219d12c722e286798b126ffed29ee1eb
wordpress callback: http://shop.home/?wc-api=WC_Gateway_Blockonomics&secret=f5e53a0f219d12c722e286798b126ffed29ee1eb
```
The current algorithm will pick one of the two server callbacks, whichever one is closer to the end of the array, and update it.

If we don't want to treat protocol and secret partial matches differently, I suggest that we only use `strpos($server_base_url, $base_url) === 0` for checking partial matches. It will catch both urls with different _protocols_ and _secrets_. Thanks to checking the position of the match it makes sure that websites with extra subdomains aren't falsely flagged. 

If we want to check for two partial matches, perhaps the smartest option is to notify the user that he has two URLs that closely resemble each other and ask him to take action; since it doesn't seem to happen often we might not want to complicate the logic for it.